tests/haskell-font-lock-tests.el

```diff
@@ -866,3 +866,55 @@
      ("Y" t haskell-constructor-face)
      ("Z" t haskell-type-face)
      ("C" t haskell-constructor-face))))
+
+(ert-deftest haskell-pattern ()
+  "Fontify the \"pattern\" keyword in contexts related to pattern synonyms."
+  :expected-result :failed
+  (check-properties
+   '("pattern A = B"
+     "pattern A <- B"
+     "pattern A ← B"
+     "pattern A n <- (subtract 1 -> n) where A n = n + 1"
+     "module Main (pattern A) where"
+     "pattern A :: a -> B"
+     "pattern A :: (C a) => a -> B"
+     "pattern A :: (C a) => a -> B"
+     "pattern A :: (C a) => () => a -> B"
+     "pattern A :: (C a) => () => a -> B"
+     "pattern A :: (C a) => (C a) => a -> B"
+     "pattern A :: (C a) => (C a) => a -> B")
+   '(("pattern"  t haskell-keyword-face)
```

Comment:
This checks only this first `pattern`. You need to write as many `patter` rules as there are `pattern` keywords in the example code. It is okay to leave out other tokens, `check-properties` skips over fragments that aren't listed.